File: data_structures/singly_linked_list/cpp/singly_linked_list.cpp
#include "singly_linked_list.h"

SinglyLinkedList::~SinglyLinkedList() {
    while (head_) { Node* n = head_->next; delete head_; head_ = n; }
}

void SinglyLinkedList::push_front(int value) {
    head_ = new Node{value, head_};
    ++size_;
}

void SinglyLinkedList::push_back(int value) {
    Node* n = new Node{value, nullptr};
    if (!head_) { head_ = n; }
    else { Node* cur = head_; while (cur->next) cur = cur->next; cur->next = n; }
    ++size_;
}

int SinglyLinkedList::pop_front() {
    if (!head_) throw std::underflow_error("list is empty");
    int v = head_->val;
    Node* n = head_->next;
    delete head_;
    head_ = n;
    --size_;
    return v;
}

int SinglyLinkedList::front() const {
    if (!head_) throw std::underflow_error("list is empty");
    return head_->val;
}

int SinglyLinkedList::back() const {
    if (!head_) throw std::underflow_error("list is empty");
    Node* cur = head_;
    while (cur->next) cur = cur->next;
    return cur->val;
}

bool SinglyLinkedList::contains(int value) const {
    for (Node* cur = head_; cur; cur = cur->next)
        if (cur->val == value) return true;
    return false;
}

bool SinglyLinkedList::remove(int value) {
    if (!head_) return false;
    if (head_->val == value) { pop_front(); return true; }
    for (Node* cur = head_; cur->next; cur = cur->next) {
        if (cur->next->val == value) {
            Node* del = cur->next;
            cur->next = del->next;
            delete del;
            --size_;
            return true;
        }
    }
    return false;
}

int SinglyLinkedList::size() const { return size_; }
bool SinglyLinkedList::is_empty() const { return size_ == 0; }

std::vector<int> SinglyLinkedList::to_vector() const {
    std::vector<int> v;
    for (Node* cur = head_; cur; cur = cur->next) v.push_back(cur->val);
    return v;
}

File: data_structures/singly_linked_list/python/singly_linked_list.py
"""単方向連結リスト

操作:
  push_front(v)       : 先頭に挿入  O(1)
  push_back(v)        : 末尾に追加  O(n)
  pop_front()         : 先頭を取り出す  O(1)
  front() / back()    : 先頭/末尾の参照  O(1)/O(n)
  contains(v)         : 値を検索  O(n)
  remove(v)           : 最初に見つかった値を削除  O(n)
  size() / is_empty() : O(1)
  to_list()           : 先頭→末尾のコピーリスト  O(n)
"""

from typing import List, Optional


from typing import Generic, TypeVar

T = TypeVar("T")


class _Node:
    __slots__ = ("val", "next")

    def __init__(self, val: T) -> None:
        self.val = val
        self.next: Optional["_Node"] = None


class SinglyLinkedList(Generic[T]):
    def __init__(self) -> None:
        self._head: Optional[_Node] = None
        self._size = 0

    def push_front(self, value: T) -> None:
        node = _Node(value)
        node.next = self._head
        self._head = node
        self._size += 1

    def push_back(self, value: T) -> None:
        node = _Node(value)
        if self._head is None:
            self._head = node
        else:
            cur = self._head
            while cur.next:
                cur = cur.next
            cur.next = node
        self._size += 1

    def pop_front(self) -> T:
        if self._head is None:
            raise IndexError("list is empty")
        v = self._head.val
        self._head = self._head.next
        self._size -= 1
        return v

    def front(self) -> T:
        if self._head is None:
            raise IndexError("list is empty")
        return self._head.val

    def back(self) -> T:
        if self._head is None:
            raise IndexError("list is empty")
        cur = self._head
        while cur.next:
            cur = cur.next
        return cur.val

    def contains(self, value: T) -> bool:
        cur = self._head
        while cur:
            if cur.val == value:
                return True
            cur = cur.next
        return False

    def remove(self, value: T) -> bool:
        if self._head is None:
            return False
        if self._head.val == value:
            self._head = self._head.next
            self._size -= 1
            return True
        cur = self._head
        while cur.next:
            if cur.next.val == value:
                cur.next = cur.next.next
                self._size -= 1
                return True
            cur = cur.next
        return False

    def size(self) -> int:
        return self._size

    def is_empty(self) -> bool:
        return self._size == 0

    def to_list(self) -> List[T]:
        result = []
        cur = self._head
        while cur:
            result.append(cur.val)
            cur = cur.next
        return result
